### PR TITLE
add types of mirrorx in test_ts

### PR DESCRIPTION
**Pull request by Aaronphy**
Please fill in this template.

- [ ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not provide its own types, and you can not add them.
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by Aaronphy** [comment]
@tdreyno @leemhenson @kamal   mirrorx depends on react but has a lower required TypeScript version.
I have no idea to solve this problem.
Please help!!!!!

File: types/mirrorx/tsconfig.json
Comment (minestarks) [inline types/mirrorx/tsconfig.json]: Please set `strictNullChecks` to `true` on all new packages.

File: types/mirrorx/index.d.ts
Comment (minestarks) [inline types/mirrorx/index.d.ts]: Are you sure you don't want `any[]` here?

File: types/mirrorx/index.d.ts
Comment (minestarks) [inline types/mirrorx/index.d.ts]: Do you need 2.4 here? Looks like react's minimum is 2.3 so you should be able to get away with 2.3?

**Review by minestarks — CHANGES_REQUESTED** [review]

File: types/mirrorx/index.d.ts
Comment (Aaronphy) [reply types/mirrorx/index.d.ts]: because  react-redux `s mininum is 2.4.

File: types/mirrorx/tsconfig.json
Comment (Aaronphy) [reply types/mirrorx/tsconfig.json]: ok.

File: types/mirrorx/index.d.ts
Comment (Aaronphy) [reply types/mirrorx/index.d.ts]: Oh,mistake. sry, I will fix it.

Comment (Aaronphy) [reply types/mirrorx/index.d.ts]: I have fixed it ,please review. Thanks.

File: types/mirrorx/tsconfig.json
Comment (Aaronphy) [reply types/mirrorx/tsconfig.json]: Done.

**Review by minestarks — APPROVED** [review]

File: types/mirrorx/tslint.json
Comment (minestarks) [inline types/mirrorx/tslint.json]: Ooops, I missed this in my last review. Please don't turn off an lint rules on new packages, they're there to do good!

**Review by minestarks — CHANGES_REQUESTED** [review]

File: types/mirrorx/tslint.json
Comment (Aaronphy) [reply types/mirrorx/tslint.json]: I have turned these rules on . Please check. Thanks. @minestarks

Comment (Aaronphy) [reply types/mirrorx/tslint.json]: some are still off ,  because people are different.

Comment (minestarks) [reply types/mirrorx/tslint.json]: The linter rules enforce the coding standards for this repo to make it easier for everybody to read and maintain these typings. Could you turn them back on before I can merge this, please? Thank you!

File: types/mirrorx/tslint.json
Comment (plantain-00) [inline types/mirrorx/tslint.json]: Means that for new packages, this file should just `"extends": "dtslint/dt.json"`, no other rules.

File: types/mirrorx/tslint.json
Comment (Aaronphy) [reply types/mirrorx/tslint.json]: OKey.

Comment (Aaronphy) [reply types/mirrorx/tslint.json]: Hi,@minestarks  I have some questions. How can I contact you? E-mail or skype? 
I install the extension "tslint" in vscode. However, it does not work. Is there any other way to auto-fix the errors ?
And the tests.ts require the module 'mirrox' in npm. 
How to add the mirrorx to the whitelist in travis?

Comment (Aaronphy) [reply types/mirrorx/tslint.json]: @minestarks  this has taken lots of my time. I just want to finish it. Please  help to solve my puzzles above!

Comment (minestarks) [reply types/mirrorx/tslint.json]: Sorry about the confusion :/

This is the best place to contact me, and others who can also help.

I don't know what you mean by the npm dependency. I'm adding a comment to your line with the `import` statement -- I think that should help?

You can look at the Travis CI log here on GitHub to see the errors. 

Or to run the linter locally, from your DefinitelyTyped repo, run:

```
npm install
npm run lint mirrorx
```

In vscode, do Alt-Shift-F to format the whole document, that will fix a lot of the whitespace errors.

Most of the other errors are fairly obvious, you can fix them line by line. I wish there was a way to auto-fix them, but I don't know of any!

File: types/mirrorx/mirrorx-tests.ts
Comment (minestarks) [inline types/mirrorx/mirrorx-tests.ts]:
```ts
import mirror = require('mirrorx');
```

Then later below, instead of `actions` you would use `mirror.actions`

**Review by minestarks — CHANGES_REQUESTED** [review]

File: types/mirrorx/mirrorx-tests.ts
Comment (Aaronphy) [reply types/mirrorx/mirrorx-tests.ts]: @minestarks sorry to disturb you. I have fixed all the errors in the lint. Please check.

Comment (Aaronphy) [reply types/mirrorx/mirrorx-tests.ts]: @minestarks Could you check it please?

**Comment by Aaronphy** [comment]
@typescript-bot When to merge this PR?

**Comment by Aaronphy** [comment]
@plantain-00 @minestarks please~~~~~check

File: notNeededPackages.json
Comment (plantain-00) [inline notNeededPackages.json]: Why `mirror` lib should be removed? it seems it don't contain types: https://github.com/mirrorjs/mirror/blob/master/package.json

**Review by plantain-00 — CHANGES_REQUESTED** [review]

File: notNeededPackages.json
Comment (Aaronphy) [reply notNeededPackages.json]: Ok， I will remove it.

Comment (Aaronphy) [reply notNeededPackages.json]: @plantain-00  Thanks, I have removed it from `notNeededPackages.json`. Please check again.

**Review by plantain-00 — APPROVED** [review]

**Comment by Aaronphy** [comment]
@plantain-00 Is it done? When can I use @types/mirrorx ?

**Comment by plantain-00** [comment]
After this is merged, there should be a few hours to be published to npm: https://github.com/DefinitelyTyped/DefinitelyTyped#my-pr-is-merged-when-will-the-types-npm-package-be-updated

**Comment by Aaronphy** [comment]
@plantain-00  However this PR is still in Changes Requested Status.  When will it be merged?

**Comment by plantain-00** [comment]
Sorry, I can't, I just do code review.

**Comment by Aaronphy** [comment]
@minestarks please merge............

**Review by aozgaa — CHANGES_REQUESTED** [review]
Just a couple questions about the `Renderer`.

Thanks for the contribution. I'll merge this in once the questions are resolved.

File: types/mirrorx/index.d.ts
Comment (Aaronphy) [reply types/mirrorx/index.d.ts]: In line 175, I have exported a render function.
```export const render: Renderer;```
```Renderer``` have many types of params and functions. So I create a ```interface``` to describe the function.

File: types/mirrorx/index.d.ts
Comment (Aaronphy) [reply types/mirrorx/index.d.ts]: No, it`s not. I will remove it.

File: types/mirrorx/index.d.ts
Comment (Aaronphy) [reply types/mirrorx/index.d.ts]: @aozgaa Thanks, please check & merge. I think the interface is better to describe the render.

Comment (Aaronphy) [reply types/mirrorx/index.d.ts]: @aozgaa @typescript-bot @dt-bot @minestarks  When to merge?

**Pull request merged**
